### DISCUSSION TITLE
How to combine Asset modules with css-loader?

### DISCUSSION THREAD

**Discussion by 4eb0da** [webpack/webpack#14067] [Q&A]
Hello
I have a project that works fine with Webpack 4 and now i want to upgrade it to Webpack 5
Right now some modules is build with `css-loader` + `extract-loader` + `raw-loader`. It used for small chunks of CSS, which is used in SSR code

What i got with Webpack 5:
`extract-loader` and `raw-loader` is deprecated by Asset modules
Okay, in simple case, i can go "asset/source" and it works fine
But, if CSS contains `url(...)`, it should be processed properly (to update the paths, extract images, etc)
With `css-loader` result asset contains intermediate module source like `// Imports\nimport ___CSS_LOADER_API_IMPORT___ from ...`, so i need something between Asset modules and css-loader
i've tried several options here:
* Use `to-string-loader`, as it suggested in docs: https://webpack.js.org/loaders/css-loader/. It doesn't work, just produces `to-string-loader`'s source. Other note: `to-string-loader` is outdated and doesn't work with ES modules from `css-loader`
* Use `extract-loader`. It is also outdated and doesn't work with current version of `css-loader`, it can't parse `import.meta.url`
* Don't use Asset modules, combine `css-loader` + `to-string-loader` + type: 'javascript/auto`. Only this works for me, but i'm very confused

What is the right way here? How can i get Asset modules + `css-loader` to work?

**Comment by alexander-akait** [comment]
`extract-loader` is deprecated and not supported, please use `mini-css-extract-plugin`

**Comment by 4eb0da** [comment]
As far as i know, `mini-css-extract-plugin` produces one CSS for each chunk with CSS
I want the other thing:
```css
// a.css
body {
    color: red;
}

// b.css
body {
    background: #fff;
}
```

```js
import stylesA from './a.css?inline';
import stylesB from './b.css?inline';

console.log(stylesA); // body { color: red; }
console.log(stylesB); // body { background: #fff; }
```

**Reply by alexander-akait** [reply]
Yes, it is expected, like future `import style from "./style.css" assert { type: "css" }` will work, so I recommend to change approach because it will be invalid in future

**Reply by 4eb0da** [reply]
I don't get it, what do you mean with `assert { type: "css" }` ? Never seen this before
Is it some kind of future syntax? If so, which way is recommended to migrate nowdays?

**Reply by alexander-akait** [reply]
https://web.dev/css-module-scripts/, it will be supported in near future using `css-loader`

**Reply by 4eb0da** [reply]
Still can't get it. Seems future syntax will work on client side and i have other use case
What approach is recommended today? Make entry for each CSS file?

**Reply by alexander-akait** [reply]
Yes, each css files is separated chunk

**Reply by 4eb0da** [reply]
Ok, understood
But this makes code more difficult to maintain
For example, in some js file i decide that i need contents of a css file
With `?inline` approach i can do it with changes just in one file
With `css in separate chunk` approach i need to change js file and webpack config (or maybe i'm missing some useful syntax to generate chunks on the fly, like dynamic import() does?)

**Reply by alexander-akait** [reply]
You can use `import()` with CSS

**Reply by 4eb0da** [reply]
Yes, and it's useful on client side
But on a server i want to do it in a sync way
I can create some weird code:
```js
const styles = fs.readFileSync('asset-output-path.css', 'utf-8');

// Not used, needs to be found by webpack to produce separate chunk, which is readden above
export function unused(){
    import('./styles.css');
}
```

compare to

```js
import styles from './styles.css?inline';
```

I still has a feeling that i'm missing something important)

**Reply by alexander-akait** [reply]
Why you need CSS on server, SSR?

**Reply by 4eb0da** [reply]
In general, if some functionality of the site is shown in rare cases (and it should be visible instantly from page load), i want to:
1) Not to include css of it into main.css
2) Place css and html on a server

**Reply by alexander-akait** [reply]
you can use multi compiler mode in this case and prepare different code for different targets, in future we will have `universal` target, but it is limited right now

**Reply by 4eb0da** [reply]
Multi compiler mode = several configurations in one config?
Do you mean that i can build `main.css` and `main_with_rare.css`?
This will work with one condition, but in my case there like dozen+ of it

**Reply by alexander-akait** [reply]
In this case your can create custom loader/plugin and implement own logic, but it is not recommended

**Reply by 4eb0da** [reply]
Okay
Big thanks for the answers!

I'm wondering some time already, if there any way to create custom "import()", which will return js / css and paths / contents to that chunk synchronously...
